packages/vitest/src/node/reporters/blob.ts

```diff
@@ -0,0 +1,115 @@
+import { mkdir, readFile, readdir, writeFile } from 'node:fs/promises'
+import { existsSync } from 'node:fs'
+import { parse, stringify } from 'flatted'
+import { dirname, resolve } from 'pathe'
+import { cleanUrl } from 'vite-node/utils'
+import type { File, Reporter, Vitest } from '../../types'
+import { getOutputFile } from '../../utils/config-helpers'
+import type { WorkspaceProject } from '../workspace'
+
+export interface BlobOptions {
+  outputFile?: string
+}
+
+export class BlobReporter implements Reporter {
+  ctx!: Vitest
+  options: BlobOptions
+
+  constructor(options: BlobOptions) {
+    this.options = options
+  }
+
+  onInit(ctx: Vitest): void {
+    if (ctx.config.watch)
+      throw new Error('Blob reporter is not supported in watch mode')
+
+    this.ctx = ctx
+  }
+
+  async onFinished(files?: File[], errors?: unknown[]) {
+    let outputFile = this.options.outputFile ?? getOutputFile(this.ctx.config, 'blob')
+    if (!outputFile) {
+      const shard = this.ctx.config.shard
+      outputFile = shard
+        ? `.vitest-reports/blob-${shard.index}-${shard.count}.json`
+        : '.vitest-reports/blob.json'
+    }
+
+    const moduleKeys = this.ctx.projects.map((project) => {
+      return [project.getName(), [...project.server.moduleGraph.idToModuleMap.keys()]]
+    })
+
+    const report = stringify([this.ctx.version, files, errors, moduleKeys])
```

Comment:
`files` and `errors` could be `undefined` based on types. After parsing it's assumed they are arrays. Let's add fallbacks. 

Also let's type the format of blob contents - it will help to spot the parts that need updating in future when new content is added there.

```suggestion
    const report = stringify([this.ctx.version, files || [], errors || [], moduleKeys] satisfies Report)
```

... where `Report` is type defined on line ~9. (TIL named entries in tuple 😮)

```ts
type Report = [
  versionNumber: string,
  files: File[],
  errors: unknown[],
  moduleKeys: (string | string[])[][],
]
```

... and then in parsing:

```ts
const [version, files, errors, moduleKeys] = parse(content) as Report
```

Reply:
Done